server/src/main/java/org/elasticsearch/persistent/AllocatedPersistentTask.java

```diff
@@ -94,35 +95,82 @@ protected final boolean isCompleted() {
         return state.get() == State.COMPLETED;
     }
 
+    protected final boolean isLocallyAborted() {
+        return state.get() == State.LOCAL_ABORTED;
+    }
+
     boolean markAsCancelled() {
         return state.compareAndSet(State.STARTED, State.PENDING_CANCEL);
     }
 
     public void markAsCompleted() {
-        completeAndNotifyIfNeeded(null);
+        completeAndNotifyIfNeeded(null, null);
     }
 
     public void markAsFailed(Exception e) {
         if (CancelTasksRequest.DEFAULT_REASON.equals(getReasonCancelled())) {
-            completeAndNotifyIfNeeded(null);
+            completeAndNotifyIfNeeded(null, null);
         } else {
-            completeAndNotifyIfNeeded(e);
+            completeAndNotifyIfNeeded(e, null);
+        }
+    }
+
+    /**
+     * Indicates that this persistent task is no longer going to run on the local node.
+     * This will cause the local task to be terminated, and the associated persistent
+     * task to be reassigned by the master node. The persistent task <em>may</em> be
+     * reassigned to the same node unless separate measures have been taken to prevent
+     * this. The task should complete any graceful shutdown actions before calling this
+     * method.
+     * @throws IllegalStateException This exception will be thrown if the cluster contains
+     *                               nodes that are too old to understand the concept of
+     *                               locally aborting tasks. In this situation callers
+     *                               should revert to whatever the functionality was prior
+     *                               to local abort being possible. It is possible to check
+     *                               if local abort is possible before starting a sequence
+     *                               of steps that will end in a local abort by calling
+     *                               {@link PersistentTasksService#isLocalAbortSupported}.
+     * @param localAbortReason Reason for the task being aborted on this node. This
+     *                         will be recorded as the reason for unassignment of the
+     *                         persistent task.
+     */
+    public void markAsLocallyAborted(String localAbortReason) {
+        if (persistentTasksService.isLocalAbortSupported() == false) {
+            throw new IllegalStateException("attempt to abort a persistent task locally in a cluster that does not support this");
```

Comment:
Can we mention the minimum node version preventing this as well as the version it is supported from? Perhaps move the throw into `persistentTasksService`.